Automattic/wp-calypso
# Checkout: Remove "onInitialize" arguments from useShoppingCartManager

#### Changes proposed in this Pull Request

In preparation to separate the shopping cart management from the main checkout page so it can be used elsewhere in calypso, we need to be able to add the products and coupons from the URL to the cart after the cart has been initially fetched.

This has presented a problem in the past because until these items have been added (and the cart refreshed), we cannot be sure that checkout is "ready", even though the cart has technically loaded. This has major implications for other parts of checkout, notably `useRedirectIfCartEmpty`, which needs to wait for this process before deciding if it should redirect (the cart may be empty when the page loads just as we're beginning to add products from the URL).

Here, we add a new hook, `useAddProductsFromUrl`, which handles the process of completing this initialization, and which keeps its own state that it uses to determine if the process has completed. Other components and hooks can then rely on this state to know if the cart is "ready" for use.

With that in place, it's possible to remove `productsToAddOnInitialize`/`couponToAddOnInitialize` from `useShoppingCartManager`, which becomes less complex and less tightly coupled to running inside checkout.

#### Testing instructions

- Visit calypso, enable record-analytics debugging by typing `localStorage.setItem('debug', 'calypso:composite-checkout:record-analytics')` into your console, and then reloading the page.
- Visit checkout with a product in the URL like `/checkout/example.com/premium`.
- Verify that checkout loads without errors and that you see the product added to your cart.
- When loading checkout, make sure that the `CART_INIT_COMPLETE` event happens in the console, that it happens after the cart is initialized, and that it only happens once.
- When loading checkout, make sure that the `CART_ADD_ITEM` event happens in your console, that it has the correct product listed, and that it happens only once.
- Remove all products from your cart.
- Verify that you are redirected to the `/plans` page.
- Visit checkout with a product and a coupon in the URL like `/checkout/example.com/premium?coupon=FOOBAR`.
- Verify that checkout loads without errors and that you see the product and the coupon added to your cart.
